### PR TITLE
fn3: MixedViewport

### PR DESCRIPTION
**Pull request by Hixie**
Also:
- Expose the slot of each Element.
- Minor improvements to HomogeneousViewport.
- Replace TestComponent with FlipComponent in tests.

### CODE REVIEW

File: sky/unit/test/fn3/homogeneous_viewport_test.dart [hunk 1]
```diff
@@ -32,39 +14,41 @@ void main() {
...
+        ),
+        right: new Text('Not Today')
+      );
     }
 
     tester.pumpFrame(builder());
 
-    StatefulComponentElement testComponent = tester.findElement((element) => element.widget is TestComponent);
-    TestComponentState testComponentState = testComponent.state;
+    FlipComponentState testComponent = tester.findElement((element) => element.widget is FlipComponent).state;
```
Comment (abarth) [inline sky/unit/test/fn3/homogeneous_viewport_test.dart]: This cause an analyzer warning

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:collection';
+
+import 'package:sky/rendering.dart';
+import 'package:sky/src/fn3/framework.dart';
+import 'package:sky/src/fn3/basic.dart';
+
+typedef Widget IndexedBuilder(int index); // return null if index is greater than index of last entry
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: You should pass a BuildContext to builder functions.

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+
+  /// Whether childOffsets includes the offset of the last child.
+  bool _didReachLastChild = false;
+
+  /// The index of the first child whose bottom edge is below the top of the
+  /// viewport.
+  int _firstVisibleChildIndex;
+
+  /// The currently visibly children.
+  Map<_ChildKey, Element> _childrenByKey = new Map<_ChildKey, Element>();
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: final

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+
+  /// The child offsets that we've been told are invalid.
+  Set<int> _invalidIndices = new Set<int>();
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: final

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+
+  RenderBlockViewport get renderObject => super.renderObject;
+
+  /// Notify the BlockViewport that the children at indices have, or might have,
+  /// changed size. Call this whenever the dimensions of a particular child
+  /// change, so that the rendering will be updated accordingly. A pointer to
+  /// this method is provided via the onInvalidatorAvailable callback.
+  void invalidate(Iterable<int> indices) {
+    _invalidIndices.addAll(indices);
+    renderObject.markNeedsLayout();
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: Should we assert that indicies is non empty?

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+  }
+
+  /// Forget all the known child offsets.
+  void _resetCache() {
+    _didReachLastChild = false;
+    _childOffsets = <double>[0.0];
+    _invalidIndices = new Set<int>();
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: _invalidIndices.clear()

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+        final _ChildKey key = new _ChildKey.fromWidget(element.widget);
+        final bool isVisible = _childOffsets[widgetIndex] < endOffset && _childOffsets[widgetIndex + 1] >= widget.startOffset;
+        if (isVisible) {
+          // Keep it.
+          newChildren[key] = element;
+          builtChildren[widgetIndex] = element;
+        } else {
+          // Drop it.
+          _childrenByKey.remove(key);
+          updateChild(element, null, null);
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: There's an easier way to do this, but this works too.

**Comment by abarth** [comment]
LGTM

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart (hunk 3 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: didn't make that one final, but i made two others final instead.

File: sky/packages/sky/lib/src/fn3/mixed_viewport.dart (hunk 7 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/mixed_viewport.dart]: what's the easier way?

**Comment by abarth** [comment]
_detachChild is the easier way.

**Pull request merged**